crates/agglayer-config/src/storage.rs

```diff
@@ -0,0 +1,153 @@
+use std::path::Path;
+use std::path::PathBuf;
+
+use serde::Deserialize;
+use serde::Serialize;
+
+const STORAGE_DIR: &str = "storage";
+const METADATA_DB_NAME: &str = "metadata";
+const PENDING_DB_NAME: &str = "pending";
+const STATE_DB_NAME: &str = "state";
+const EPOCHS_DB_PATH: &str = "epochs";
+
+/// Configuration for the storage.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct StorageConfig {
+    /// Custom metadata storage path or inferred from the db path.
+    pub metadata_db_path: PathBuf,
+    /// Custom pending storage path or inferred from the db path.
+    pub pending_db_path: PathBuf,
+    /// Custom state storage path or inferred from the db path.
+    pub state_db_path: PathBuf,
+    /// Custom epochs storage path or inferred from the db path.
+    pub epochs_db_path: PathBuf,
+}
+
+impl StorageConfig {
+    pub fn path_contextualized(mut self, base_path: &Path) -> Self {
+        if !self.metadata_db_path.is_absolute() {
+            self.metadata_db_path = base_path.join(&self.metadata_db_path);
+        }
+
+        if !self.pending_db_path.is_absolute() {
+            self.pending_db_path = base_path.join(&self.pending_db_path);
+        }
+
+        if !self.state_db_path.is_absolute() {
+            self.state_db_path = base_path.join(&self.state_db_path);
+        }
+
+        if !self.epochs_db_path.is_absolute() {
+            self.epochs_db_path = base_path.join(&self.epochs_db_path);
+        }
+
+        self
+    }
+
+    /// Creates a new storage configuration with the default path.
+    pub fn new_with_default_path() -> Self {
+        let base_path: PathBuf = Path::new("./").join(STORAGE_DIR);
+
+        Self {
+            metadata_db_path: default_metadata_path(&base_path),
+            pending_db_path: default_pending_path(&base_path),
+            state_db_path: default_state_path(&base_path),
+            epochs_db_path: default_epochs_path(&base_path),
+        }
+    }
+}
+
+impl From<&Path> for StorageConfig {
+    fn from(value: &Path) -> Self {
+        let base = value.join(STORAGE_DIR);
+
+        Self {
+            metadata_db_path: base.join(METADATA_DB_NAME),
+            pending_db_path: base.join(PENDING_DB_NAME),
+            state_db_path: base.join(STATE_DB_NAME),
+            epochs_db_path: base.join(EPOCHS_DB_PATH),
+        }
+    }
+}
```

Comment:
Having this as a `From` impl seems a bit too magical. Would prefer an explicit constructor, like `StorageConfig::from_base_path` for this.